### ISSUE TITLE
[Bug]: Cannot use aws_elasticache_user_group_association to specify the default user

### ISSUE THREAD

**Issue by CarlosZaldivar** [hashicorp/terraform-provider-aws#34784]
### Terraform Core Version

1.4.6

### AWS Provider Version

4.67.0

### Affected Resource(s)

- aws_elasticache_user_group
- aws_elasticache_user_group_association

### Expected Behavior

I expect to be able to add a default user to `aws_elasticache_user_group` using `aws_elasticache_user_group_association`, not just `aws_elasticache_user_group.user_ids`. The usage of `aws_elasticache_user_group.user_ids` conflicts with adding new users using `aws_elasticache_user_group_association` and forces the use of `ignore_changes` on `user_ids`. This has been described [here](https://github.com/hashicorp/terraform-provider-aws/issues/22870#issuecomment-1097324435).

### Actual Behavior

Error 400 is shown.

### Relevant Error/Panic Output Snippet

```shell
Error: creating ElastiCache User Group (test-group): DefaultUserRequired: User group needs to contain a user with the user name default.
```

### Terraform Configuration Files

```terraform
resource "aws_elasticache_user" "test" {
  user_id       = "test-user"
  user_name     = "default"
  access_string = "on ~* &* +@all"
  engine        = "REDIS"
  passwords     = ["testtesttesttesttest"]
}

resource "aws_elasticache_user_group" "test_group" {
  engine        = "REDIS"
  user_group_id = "test-group"
}

resource "aws_elasticache_user_group_association" "test_user_association" {
  user_group_id = aws_elasticache_user_group.test_group.user_group_id
  user_id       = aws_elasticache_user.test.id
}
```

### Steps to Reproduce

Run apply

### Would you like to implement a fix?

None

**Comment by Sohaib112** [comment]
I believe this limitation is from AWS. I tried creating a user group using the AWS console and it gave me a similar error to add the `default` user in user_group.
I did not find any AWS document explaining this.

**Comment by CarlosZaldivar** [comment]
Yes, but could the provider be smart enough to add the default user from `aws_elasticache_user_group_association` when creating the group? As far as I know `aws_elasticache_user_group_association` is a concept from the provider and it doesn't have an AWS equivalent anyway. I don't know the inner workings of the provider, so I don't know how difficult it would be to make it work.

**Comment by OJFord** [comment]
It is not, or is not any longer, an AWS limitation - having created the group with terraform I have removed the `"default"` user in AWS console no problem.

**Comment by sukumaar** [comment]
Hello @CarlosZaldivar,

Thank you for raising this. I attempted to reproduce this and here are my findings.

This is an AWS API limitation - AWS requires a user with `user_name = "default"` to be present at user group creation time. The `aws_elasticache_user_group_association` resource fires *after* group creation, so it cannot satisfy this requirement.

AWS documentation reference:
- [RBAC documentation](https://docs.aws.amazon.com/AmazonElastiCache/latest/dg/Clusters.RBAC.html) states: *"ElastiCache automatically configures a default user with user ID and user name "default" and adds it to all user groups. You can't modify or delete this user."*
- [CloudFormation UserGroup documentation](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-elasticache-usergroup.html) states: *"A user named default must be included."*

> could the provider be smart enough to add the default user from `aws_elasticache_user_group_association` when creating the group?

I understand the ask, but this would go against Terraform's [Provider Design Principles](https://developer.hashicorp.com/terraform/plugin/best-practices/hashicorp-provider-design-principles):

- **"Resource and attribute schema should closely match the underlying API"** - the provider should reflect what the API expects, not silently add parameters behind the scenes.

The solution is to include the default user in `user_ids` at group creation, use `ignore_changes` on `user_ids` to prevent drift, and manage additional users via `aws_elasticache_user_group_association`:

```hcl
resource "aws_elasticache_user" "default" {
  user_id       = "my-default-user"
  user_name     = "default"
  access_string = "on ~* &* +@all"
  engine        = "redis"
  passwords     = ["testtesttesttesttest"]
}

resource "aws_elasticache_user" "app_user" {
  user_id       = "my-app-user"
  user_name     = "appuser"
  access_string = "on ~app::* -@all +@read +@write"
  engine        = "redis"
  passwords     = ["anotherstrongpassword1"]
}

resource "aws_elasticache_user_group" "example" {
  engine        = "redis"
  user_group_id = "my-user-group"
  user_ids      = [aws_elasticache_user.default.user_id]

  lifecycle {
    ignore_changes = [user_ids]
  }
}

resource "aws_elasticache_user_group_association" "app_user" {
  user_group_id = aws_elasticache_user_group.example.user_group_id
  user_id       = aws_elasticache_user.app_user.user_id
}
```

Also worth noting - if you are using Valkey engine, there is no default user requirement. The [RBAC documentation](https://docs.aws.amazon.com/AmazonElastiCache/latest/dg/Clusters.RBAC.html) explicitly states: *"There is no default user requirement. When the Valkey user group is attached to clusters, the default user requirement is automatically disabled."*

This is the terraform version I was using:

```console
$ terraform -v
Terraform v1.14.5
on darwin_arm64
+ provider registry.terraform.io/hashicorp/aws v6.46.0

Your version of Terraform is out of date! The latest version
is 1.15.4. You can update by downloading from https://developer.hashicorp.com/terraform/install
```

---
> [!NOTE]
> Not a bug, please feel free to close this issue. The `upstream` label is appropriate here - this is an AWS API constraint, and the provider correctly reflects it.
